### PR TITLE
Add cached array.ts

### PR DESCRIPTION
**Pull request by minggo**
@dumganhar

### CODE REVIEW

File: cocos/core/memop/CachedArray.h [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+public:
+    using CompareFn = std::function<int(T *, T *)>;
+
+    CachedArray(uint32_t length);
+
+    /**
+     * @param length Initial length
+     * @param compareFn Comparison function for sorting
+     */
+    CachedArray(uint32_t length, CompareFn compareFn);
```
Comment (dumganhar) [inline cocos/core/memop/CachedArray.h:52]: CompareFn -> const CompareFn&

File: cocos/core/memop/CachedArray.h [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+ * Cached array is a data structure for objects cache, it's designed for persistent data.
+ * Its content array length will keep grow.
+ * @zh
+ * 适用于对象缓存的数组类型封装，一般用于不易被移除的常驻数据。
+ * 它的内部数组长度会持续增长，不会减少。
+ */
+template <typename T>
+class CachedArray {
+public:
+    using CompareFn = std::function<int(T *, T *)>;
```
Comment (dumganhar) [inline cocos/core/memop/CachedArray.h:44]: int -> int32_t

File: cocos/core/memop/CachedArray.h [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    void sort();
+
+    /**
+     * @en
+     * Add all elements of a given array to the end of the current array
+     * @zh
+     * 添加一个指定数组中的所有元素到当前数组末尾
+     * @param array The given array to be appended
+     */
+    void concat(const std::vector<T *> &);
```
Comment (dumganhar) [inline cocos/core/memop/CachedArray.h:113]: 参数名称漏了？

File: cocos/core/memop/CachedArray.h [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+     * @param idx The index of the element to be deleted
+     */
+    void fastRemove(uint32_t idx);
+
+    /**
+     * @en Returns the first index at which a given element can be found in the array.
+     * @zh 返回在数组中可以找到一个给定元素的第一个索引。
+     * @param val The element
+     */
+    int32_t indexOf(T *);
```
Comment (dumganhar) [inline cocos/core/memop/CachedArray.h:127]: 参数名称？

**Comment by minggo** [comment]
在 https://github.com/minggo/engine-native/pull/26 解决

**Pull request closed without merging**